File: nim_core/android/src/main/kotlin/com/netease/nimflutter/services/FLTAudioRecorderService.kt
/*
 * Copyright (c) 2022 NetEase, Inc. All rights reserved.
 * Use of this source code is governed by a MIT license that can be
 * found in the LICENSE file.
 */

package com.netease.nimflutter.services

import android.content.Context
import com.netease.nimflutter.FLTService
import com.netease.nimflutter.NimCore
import com.netease.nimflutter.NimResult
import com.netease.nimlib.sdk.media.record.AudioRecorder
import com.netease.nimlib.sdk.media.record.IAudioRecordCallback
import com.netease.nimlib.sdk.media.record.RecordType
import java.io.File

class FLTAudioRecorderService(
    applicationContext: Context,
    nimCore: NimCore
) : FLTService(applicationContext, nimCore) {

    private var recordType: RecordType = RecordType.AAC
    private var maxLength: Int = AudioRecorder.DEFAULT_MAX_AUDIO_RECORD_TIME_SECOND

    private var mAudioRecordCallback: IAudioRecordCallback = object : IAudioRecordCallback {
        override fun onRecordReady() {
            notify("onRecordReady", mutableMapOf())
        }

        override fun onRecordStart(audioFile: File?, recordType: RecordType?) {
            val messageMap = mutableMapOf<String, Any?>()
            messageMap["filePath"] = audioFile?.path
            messageMap["recordType"] = recordType?.fileSuffix
            notify("onRecordStart", messageMap)
        }

        override fun onRecordSuccess(audioFile: File?, audioLength: Long, recordType: RecordType?) {
            val messageMap = mutableMapOf<String, Any?>()
            messageMap["filePath"] = audioFile?.path
            messageMap["recordType"] = recordType?.fileSuffix
            messageMap["fileSize"] = audioFile?.length()
            messageMap["duration"] = audioLength
            notify("onRecordSuccess", messageMap)
        }

        override fun onRecordFail() {
            notify("onRecordFail", mutableMapOf())
        }

        override fun onRecordCancel() {
            notify("onRecordCancel", mutableMapOf())
        }

        override fun onRecordReachedMaxTime(maxTime: Int) {
            val messageMap = mutableMapOf<String, Any?>()
            messageMap["maxDuration"] = maxTime
            notify("onRecordReachedMaxTime", mutableMapOf())
            mAudioRecorder.handleEndRecord(true, maxTime)
        }
    }
    private var mAudioRecorder: AudioRecorder = AudioRecorder(
        applicationContext,
        recordType,
        maxLength,
        mAudioRecordCallback
    )

    override val serviceName = "AudioRecorderService"

    init {
        registerFlutterMethodCalls(
            "startRecord" to ::startAudioRecord,
            "stopRecord" to ::stopAudioRecord,
            "cancelRecord" to ::cancelAudioRecord,
            "isAudioRecording" to ::isAudioRecord,
            "getAmplitude" to ::getCurrentRecordAmplitude
        )
    }

    // 通知事件，将播放的监听的状态信息，通过该方法回调给Flutter层
    private fun notify(state: String, arguments: MutableMap<String, Any?>) {
        arguments["recordState"] = state
        notifyEvent("onRecordStateChange", arguments)
    }

    private suspend fun startAudioRecord(arguments: Map<String, *>): NimResult<Boolean> {
        // 如果正在录制，需要停止
        if (mAudioRecorder.isRecording) {
            mAudioRecorder.completeRecord(true)
        }
        var type = recordType
        var maxLen = maxLength
        if (arguments["recordType"] != null) {
            type =
                if ((arguments["recordType"] as Number).toInt() == 0) RecordType.AAC else RecordType.AMR
        }
        if (arguments["maxLength"] != null) {
            maxLen = (arguments["maxLength"] as Number).toInt()
        }
        // 配置参数发送变化，需要重新创建AudioRecord
        if (type != recordType || maxLen != maxLength) {
            mAudioRecorder.destroyAudioRecorder()
            recordType = type
            maxLength = maxLen
            mAudioRecorder = AudioRecorder(
                applicationContext,
                recordType,
                maxLength,
                mAudioRecordCallback
            )
        }
        mAudioRecorder.startRecord()
        return NimResult(code = 0, data = true)
    }

    @Suppress("UNUSED_PARAMETER")
    private suspend fun stopAudioRecord(arguments: Map<String, *>): NimResult<Nothing> {
        mAudioRecorder.completeRecord(false)
        return NimResult.SUCCESS
    }

    @Suppress("UNUSED_PARAMETER")
    private suspend fun cancelAudioRecord(arguments: Map<String, *>): NimResult<Nothing> {
        mAudioRecorder.completeRecord(true)
        return NimResult.SUCCESS
    }

    @Suppress("UNUSED_PARAMETER")
    private suspend fun isAudioRecord(arguments: Map<String, *>): NimResult<Boolean> =
        NimResult(code = 0, data = mAudioRecorder.isRecording)

    @Suppress("UNUSED_PARAMETER")
    private suspend fun getCurrentRecordAmplitude(arguments: Map<String, *>): NimResult<Int> =
        NimResult(code = 0, data = mAudioRecorder.currentRecordMaxAmplitude)
}

File: nim_core_v2/ios/Classes/services/FLTAudioRecorderService.swift
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import Foundation
import NIMSDK

enum AudioRecorderType: String {
  case StartRecord = "startRecord"
  case StopRecord = "stopRecord"
  case CancelRecord = "cancelRecord"
  case AudioRecording = "isAudioRecording"
  case Amplitude = "getAmplitude"
}

class FLTAudioRecorderService: FLTBaseService, FLTService {
  lazy var mediaManager = NIMSDK.shared().mediaManager
  private var duration: Double = 0
  private var maxDuration: Double = 100

  func serviceName() -> String {
    ServiceType.AudioRecordService.rawValue
  }

  override func onInitialized() {
    NIMSDK.shared().mediaManager.add(self)
  }

  deinit {
    NIMSDK.shared().mediaManager.remove(self)
  }

  func onMethodCalled(_ method: String, _ arguments: [String: Any],
                      _ resultCallback: ResultCallback) {
    switch method {
    case AudioRecorderType.StartRecord.rawValue:
      startAudioRecord(arguments, resultCallback)
    case AudioRecorderType.StopRecord.rawValue:
      stopAudioRecord(arguments, resultCallback)
    case AudioRecorderType.CancelRecord.rawValue:
      cancelAudioRecord(arguments, resultCallback)
    case AudioRecorderType.AudioRecording.rawValue:
      isAudioRecord(arguments, resultCallback)
    case AudioRecorderType.Amplitude.rawValue:
      getCurrentRecordAmplitude(arguments, resultCallback)
    default:
      break
    }
  }

  func register(_ nimCore: NimCore) {
    self.nimCore = nimCore
    nimCore.addService(self)
  }

  func startAudioRecord(_ argument: [String: Any], _ result: ResultCallback) {
    duration = 0
    maxDuration = getDuration(argument)
    mediaManager.record(getRecordAudioType(argument), duration: maxDuration)
    successCallBack(result, true)
  }

  func stopAudioRecord(_ argument: [String: Any], _ result: ResultCallback) {
    mediaManager.stopRecord()
    successCallBack(result, true)
  }

  func cancelAudioRecord(_ argument: [String: Any], _ result: ResultCallback) {
    mediaManager.cancelRecord()
    successCallBack(result, true)
  }

  func isAudioRecord(_ argument: [String: Any], _ result: ResultCallback) {
    let success = mediaManager.isRecording()
    successCallBack(result, success)
  }

  func getCurrentRecordAmplitude(_ argument: [String: Any], _ result: ResultCallback) {
    let volume = Int(mediaManager.recordPeakPower())
    successCallBack(result, volume)
  }
}

extension FLTAudioRecorderService: NIMMediaManagerDelegate {
  func recordAudio(_ filePath: String?, didBeganWithError error: Error?) {
    var arguments = [String: Any]()
    arguments["filePath"] = filePath
    arguments["maxDuration"] = Int(maxDuration)
    if filePath != nil {
      if filePath!.hasSuffix("amr") {
        arguments["recordType"] = ".amr"
      } else if filePath!.hasSuffix("aac") {
        arguments["recordType"] = ".acc"
      }
    }
    notifyEvent("onRecordStart", &arguments)
  }

  func recordAudio(_ filePath: String?, didCompletedWithError error: Error?) {
    var arguments = [String: Any]()
    arguments["filePath"] = filePath
    arguments["duration"] = Int(duration)
    arguments["maxDuration"] = Int(maxDuration)
    if filePath != nil {
      do {
        let attr = try FileManager.default.attributesOfItem(atPath: filePath!)
        let fileSize = attr[FileAttributeKey.size] as? UInt64
        arguments["fileSize"] = fileSize
      } catch {
        print("Error: \(error)")
      }
      if filePath!.hasSuffix("amr") {
        arguments["recordType"] = ".amr"
      } else if filePath!.hasSuffix("aac") {
        arguments["recordType"] = ".acc"
      }
    }
    notifyEvent("onRecordSuccess", &arguments)
    duration = 0
  }

  func recordAudioProgress(_ currentTime: TimeInterval) {
    // 对齐安卓返回毫秒
    duration = currentTime * 1000
  }

  func recordAudioDidCancelled() {
    var arguments = [String: Any]()
    notifyEvent("onRecordCancel", &arguments)
  }

  func notifyEvent(_ state: String, _ arguments: inout [String: Any]) {
    arguments["serviceName"] = serviceName()
    arguments["recordState"] = state
    nimCore?.getMethodChannel()?.invokeMethod("onRecordStateChange", arguments)
  }
}
